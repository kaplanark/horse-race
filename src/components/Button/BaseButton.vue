<script setup>
const props = defineProps({
	name: {
		type: String,
		default: 'Button'
	},
	variant: {
		type: String,
		default: 'primary'
	},
	disabled: {
		type: Boolean,
		default: false
	}
});
</script>

<template>
	<button :class="`btn btn--${props.variant}`" :disabled="props.disabled">
		{{ props.name }}
	</button>
</template>

<style lang="scss">
.btn {
	cursor: pointer;
	outline: none;
	border: none;
	z-index: 1;
	position: relative;
	font-size: inherit;
	font-family: inherit;

	&:disabled {
		cursor: default;
		opacity: 0.5;

		&:hover {
			color: var(--color-primary);

			&::before {
				display: none;
			}
		}
	}

	&--primary {
		color: var(--color-primary);
		padding: 8px 16px;
		outline: none;
		border: none;
		background-color: var(--color-white);
		transition: all 0.25s ease-in-out;
		border: 1px solid var(--color-primary);
		border-radius: 4px;
		overflow: hidden;

		&::before {
			content: '';
			z-index: -1;
			position: absolute;
			top: 0;
			bottom: 0;
			left: 0;
			right: 0;
			background-color: var(--color-primary);
			transform-origin: center right;
			transform: scaleX(0);
			transition: transform 0.25s ease-in-out;
		}

		&:hover {
			color: var(--color-white);

			&::before {
				transform-origin: center left;
				transform: scaleX(1);
			}
		}

		&:active,
		&:focus {
			color: var(--color-white);

			&::before {
				transform: scale(1);
				background-color: var(--color-primary);
			}
		}
	}

	&--secondary {
		color: var(--color-secondary);
		padding: 8px 16px;
		outline: none;
		border: none;
		background-color: var(--color-white);
		transition: all 0.25s ease-in-out;
		border: 1px solid var(--color-secondary);
		border-radius: 4px;
		overflow: hidden;

		&::before {
			content: '';
			z-index: -1;
			position: absolute;
			top: 0;
			bottom: 0;
			left: 0;
			right: 0;
			background-color: var(--color-secondary);
			transform-origin: center right;
			transform: scaleX(0);
			transition: transform 0.25s ease-in-out;
		}

		&:hover {
			color: var(--color-white);

			&::before {
				transform-origin: center left;
				transform: scaleX(1);
			}
		}

		&:active,
		&:focus {
			color: var(--color-white);

			&::before {
				transform: scale(1);
				background-color: var(--color-secondary);
			}
		}
	}
}
</style>