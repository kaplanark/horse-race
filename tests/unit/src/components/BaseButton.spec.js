import { describe, expect, test } from 'vitest';
import { mount } from '@vue/test-utils';
import BaseButton from '@components/Button/BaseButton.vue';

describe('BaseButton.vue', () => {
	let wrapper;
	beforeEach(() => {
		wrapper = mount(BaseButton, {
			props: {
				name: 'Button',
				variant: 'primary',
				disbaled: false
			}
		});
	});

	afterEach(() => {
		wrapper.unmount();
	});

	test('is correctly rendered', () => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	test('should renders is page content is correct', () => {
		expect(wrapper.vm).toBeTruthy();
	});

	test('renders props.name when passed', () => {
		expect(wrapper.text()).toContain('Button');
	});

	test('renders props.variant when passed', () => {
		expect(wrapper.classes()).toContain('btn--primary');
	});

	test('renders props.disabled when passed', () => {
		expect(wrapper.attributes('disabled')).toBeUndefined();
	});
});