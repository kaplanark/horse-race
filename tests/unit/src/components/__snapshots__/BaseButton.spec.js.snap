// Vitest Snapshot v1

exports[`BaseButton.vue > is correctly rendered 1`] = `"<button class=\\"btn btn--primary\\" disbaled=\\"false\\">Button</button>"`;
